### ISSUE TITLE
"Not a LE executable" with a MS-DOS exe

### ISSUE THREAD

**Issue by neuromancer** [samunders-core/le_disasm#3]
I'm trying to use the latest revision of le_disasm with the [Threat MS-DOS game](https://www.dosgamesarchive.com/download/threat/), however, the executable header is not correctly parsed:

```
$ ./le_disasm THREAT.EXE 
stack trace:
  ./le_disasm : Error::Error()+0x5d
  ./le_disasm : Header::throwOnInvalidSignature(std::istream&, unsigned int&)+0x1db
  ./le_disasm : Header::Header(std::istream&, unsigned int&)+0x52
  ./le_disasm : LinearExecutable::LinearExecutable(std::istream&, unsigned int)+0x3a
  ./le_disasm : main()+0xba
  /usr/lib/libc.so.6 : __libc_start_main()+0xf3
  ./le_disasm : _start()+0x2e
Not a LE executable
```

**Comment by samunders-core** [comment]
Yes, that part is not finished, feel free to send PR. Right now feeding it extender-striped version of exe file (starting with LE) should get you the furthest.

**Comment by klei1984** [comment]
Hello @neuromancer , it seems that Threat is a 16-bit MS-DOS program and loads directly into the 640 kilobytes conventional memory available to DOS.
 The error message above correctly indicates that the analyzed executable does not have a linear executable header. There is no extender to be unboud here. Either an MS-DOS load module needs to be implemented/integrated into le_disasm or an alternative would be to run the game in dosbox debugger and dump the loaded application from RAM at run-time to a flat binary image that could be fed into the analyzer of le_disasm with less coding effort.

**Comment by neuromancer** [comment]
Uhm, interesting. Maybe is it like the [Rayman executable](https://github.com/marespiaut/rayman_disasm)?

**Comment by klei1984** [comment]
Nope. Threat is a simple MZ executable. Rayman had PMODE/W extender which additionally compressed the extended 32-bit linear executable.
16-bit MS-DOS (MZ) executables can only handle up to 640 kilobytes of RAM. To enable utilization of more memory DOS extenders were invented. But an extender was not required for this game.
In threat.doc we can find the following system requirements:

Requirements :
ßßßßßßßßßßßß
   - 80286 or better                    ( 386 recommended )
   - 600,000 bytes memory with sounds   ( 550,000 without sounds )
   - 256k VGA
   - DOS 3.3 or newer
   - Hard disk                          ( about 1.5 Megas )

As you can see 600 kiB of RAM is required for the game. It does not require a DOS extender to access more than 640 kilobytes.

You can try to implement an MZ image loader within le_disasm based on the MZ exe format specification found here : http://www.delorie.com/djgpp/doc/exe/ and then could try to disassemble the 16-bit relocated flat executable image contents in 16-bit mode.
Problem is that the entire resulting code base will be 16-bit 386p assembly that modern systems do not support without emulation. Applications built for DOS extenders mainly wrap around mixed 16/32-bit linear executables with emphasis on the 32-bit part. If there are small 16-bit segments within the linear executable image they typically implement HW specific interrupt vectors and similar low level stuff while the majority of the code is 32-bits that could potentially work on modern systems as well with way less reengineering effort.
Maybe you could also try Boomerang which emits C code instead of assembler: https://github.com/BoomerangDecompiler/boomerang . Unfortunately I cannot compile Boomerang on my system and I do not want to upgrade my OS just to try this tool out...

Hope this helps.
